core/src/main/java/hudson/tasks/ArtifactArchiver.java

```diff
@@ -233,6 +251,19 @@ public void perform(Run<?,?> build, FilePath ws, Launcher launcher, TaskListener
 
             Map<String,String> files = ws.act(new ListFiles(artifacts, excludes, defaultExcludes, caseSensitive));
             if (!files.isEmpty()) {
+                if (isFlattenDirectories()) {
+                    Map<String,String> flattenedFiles = new LinkedHashMap<>();
+                    for (Map.Entry<String,String> soloFile : files.entrySet()) {
+                        String simpleFile = soloFile.getKey().replaceFirst(".+/", "");
+                        if (flattenedFiles.put(simpleFile, soloFile.getKey()) != null) {
+                            throw new AbortException("Attempted to add duplicate filenames to flattened archive. " +
```

Comment:
Seems to me this should be configurable (e.g. use deterministic first, use deterministic last, abort)